google/cloud/spanner/row.h

```diff
@@ -240,6 +240,23 @@ class Row {
   }
   ///@}
 
+  /**
+   * Allows Google Test to print internal debugging information when test
+   * assertions fail.
+   *
+   * @warning This is intended for debugging and human consumption only, not
+   *   machine consumption as the output format may change without notice.
+   */
+  friend void PrintTo(Row const& r, std::ostream* os) {
+    char const* sep = "{";
+    for (auto const& v : r.values()) {
+      *os << sep;
+      PrintTo(v, os);
```

Comment:
It may be good to use `testing::PrintToString` here, which will work on types that don't have `op<<` or `PrintTo` (which I suspect is the case with like, vector and tuple). See https://github.com/google/googletest/blob/master/googletest/docs/advanced.md

But I'd also like to avoid including a googletest header in this public .h file. Maybe we could put this `PrintTo` body in a .cc file?

What do you think?

Reply:
> It may be good to use `testing::PrintToString` here, which will work on types that don't have `op<<` or `PrintTo` (which I suspect is the case with like, vector and tuple).

The `v` are always a `spanner::Value`, right?  So I am actually calling `google::cloud::spanner::PrintTo(Value const&, std::ostream*)` and we know `Value` has a `PrintTo`.

> But I'd also like to avoid including a googletest header in this public .h file. Maybe we could put this `PrintTo` body in a .cc file?

That would be hard because I need a `PrintTo(Row<Ts...> const& r, std::ostream* os)`, it does not look like a template here because it is in the middle of the template class definition.

Reply:
My mistake. You're right. I overlooked the fact that you're calling `.values()`. What you have now LGTM.